### PR TITLE
Sync polls rules (PSG-77, PSG-1097)

### PR DESCRIPTION
**Pull request by alfogrillo**
### Description

This PR add completion blocks for updating push rules.
Completion blocks have been added for:
- Enabling/disabling a push rule
- Updated actions of a push rule

### CODE REVIEW

File: MatrixSDK/NotificationCenter/MXNotificationCenter.h [hunk 1]
```diff
@@ -197,8 +197,9 @@ extern NSString *const kMXNotificationCenterAllOtherRoomMessagesRuleID;
  
  @param pushRule the push rule to update
  @param enable YES to enable.
+ @param completion an optional completion block for the operation.
  */
-- (void)enableRule:(MXPushRule*)pushRule isEnabled:(BOOL)enable;
+- (void)enableRule:(MXPushRule*)pushRule isEnabled:(BOOL)enable completion:(void (^_Nullable)(NSError * _Nullable error))completion;
```
Comment (stefanceriu) [inline MatrixSDK/NotificationCenter/MXNotificationCenter.h:202]: If I'm not mistaken adding nullability specifiers for just parts of this file will introduce a ton of new warnings for the rest of the items. You should probably have a look at `MX_ASSUME_MISSING_NULLABILITY_BEGIN`

File: MatrixSDK/NotificationCenter/MXNotificationCenter.h [hunk 2]
```diff
@@ -208,12 +209,14 @@ extern NSString *const kMXNotificationCenterAllOtherRoomMessagesRuleID;
...
  @param highlight enable/disable highlight option.
+ @param completion an optional completion block for the operation.
  */
 - (void)updatePushRuleActions:(NSString*)ruleId
                          kind:(MXPushRuleKind)kind
                        notify:(BOOL)notify
                     soundName:(NSString*)soundName
-                    highlight:(BOOL)highlight;
+                    highlight:(BOOL)highlight
+                   completion: (void (^_Nullable)(NSError * _Nullable error))completion;
```
Comment (stefanceriu) [inline MatrixSDK/NotificationCenter/MXNotificationCenter.h:219]: I personally find the `completion: (nullable void (^)(NSError * _Nullable error))completion;` version easier to prse

File: MatrixSDK/NotificationCenter/MXNotificationCenter.m [hunk 3]
```diff
@@ -450,12 +456,21 @@ - (void)updatePushRuleActions:(NSString*)ruleId
...
+                completion(nil);
+            }
         } failure:^(NSError *error) {
             [[NSNotificationCenter defaultCenter] postNotificationName:kMXNotificationCenterDidFailRulesUpdate object:self userInfo:@{kMXNotificationCenterErrorKey:error}];
+            if (completion) {
+                completion(nil);
+            }
         }];
     }
                                                  failure:^(NSError *error) {
```
Comment (stefanceriu) [inline MatrixSDK/NotificationCenter/MXNotificationCenter.m:469]: Let's maybe move this failure to the previous line

**Review by stefanceriu — APPROVED** [review]
Some comments inline, looks good otherwise 👍

File: MatrixSDK/NotificationCenter/MXNotificationCenter.h (hunk 1 above)
Comment (alfogrillo) [reply MatrixSDK/NotificationCenter/MXNotificationCenter.h]: Oh, I didn't know.
Applied -> https://github.com/matrix-org/matrix-ios-sdk/pull/1702/commits/1f5997c8aee6cc36bccbd1b3ac1cbc37c020aca2

File: MatrixSDK/NotificationCenter/MXNotificationCenter.h (hunk 2 above)
Comment (alfogrillo) [reply MatrixSDK/NotificationCenter/MXNotificationCenter.h]: Yeah, it looks better tbh!
https://github.com/matrix-org/matrix-ios-sdk/pull/1702/commits/1f5997c8aee6cc36bccbd1b3ac1cbc37c020aca2

File: MatrixSDK/NotificationCenter/MXNotificationCenter.m (hunk 3 above)
Comment (alfogrillo) [reply MatrixSDK/NotificationCenter/MXNotificationCenter.m:469]: Can you clarify this suggestion?
Rationale so far: call the completion block **without** an error after `refreshRules` completes (even if there's an error on that).
This is because the error I'm caring of here is related to the update of the `actions` and not to the get of the push rules.

Comment (stefanceriu) [reply MatrixSDK/NotificationCenter/MXNotificationCenter.m:469]: Ah no sorry, I was just talking about alignment:
Screenshot 2023-01-31 at 10 28 59

vs. 

Screenshot 2023-01-31 at 10 29 04

Comment (alfogrillo) [reply MatrixSDK/NotificationCenter/MXNotificationCenter.m:469]: The alignment looks ok to me.
This is what I get doing ctrl+I on the full method (as it's now).

align

wdyt?

Comment (stefanceriu) [reply MatrixSDK/NotificationCenter/MXNotificationCenter.m:469]: I just find it weird because that failure block belongs to the parent method but it's indented way deep into the hierarchy, making it hard to parse. I much rather prefer this version:

Screenshot 2023-01-31 at 11 18 31

But in the grand scheme of things, this doesn't really matter. I'll leave it completely up to you 👍

Comment (alfogrillo) [reply MatrixSDK/NotificationCenter/MXNotificationCenter.m]: Applied ;)
https://github.com/matrix-org/matrix-ios-sdk/pull/1702/commits/0fe0d740256f71793e44c3fa04ba0a7f9638543c
